File: ipscromp.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>
#include <stdarg.h>
#include <termios.h>
#include <netdb.h>
#include <ctype.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "common.h"

#define DEFAULT_HOST "labrat.lexington.ibm.com"
#define DEFAULT_PORT 2002
#define DEFAULT_SERVICE "ipscromp"

char *ip_string(char *data)
{
	struct sockaddr_storage ss;
	socklen_t sslen;

	if (string_to_sockaddr(data, &ss, &sslen) < 0) {
		return NULL;
	}

	return sockaddr_to_string(&ss, sslen);
}


int connect_host(char *host, int port)
{
	int fd;
	struct addrinfo hints, *res, *rp;
	char port_str[16];
	int rc;

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_UNSPEC;     /* Allow IPv4 or IPv6 */
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_flags = AI_ADDRCONFIG;

	snprintf(port_str, sizeof(port_str), "%d", port);

	rc = getaddrinfo(host, port_str, &hints, &res);
	if (rc != 0) {
		fprintf(stderr, "Unable to determine address of '%s': %s\n",
				host, gai_strerror(rc));
		return -1;
	}

	/* Try each address until we successfully connect */
	for (rp = res; rp != NULL; rp = rp->ai_next) {
		fd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
		if (fd < 0) {
			continue;
		}

		if (debug > 1) {
			struct sockaddr_storage ss;
			memcpy(&ss, rp->ai_addr, rp->ai_addrlen);
			char *addr_str = sockaddr_to_string(&ss, rp->ai_addrlen);
			dbg("connect_host() trying %s:%d\n", addr_str ? addr_str : "unknown", port);
			free(addr_str);
		}

		if (connect(fd, rp->ai_addr, rp->ai_addrlen) == 0) {
			/* Success */
			freeaddrinfo(res);
			return fd;
		}

		close(fd);
	}

	/* All connection attempts failed */
	freeaddrinfo(res);
	perror("connect()");
	return -4;
}


int find_port(char *host, char *default_service, int default_port)
{
	struct servent *se;
	int port = -1;
	char *preferred = NULL;

	if (host != NULL && (preferred = strrchr(host, ':')) != NULL) {
		*preferred = '\0';
		preferred++;
	}

	if (preferred != NULL && (se = getservbyname(preferred, "tcp")) != NULL) {
		port = ntohs(se->s_port);
	}
	else if (preferred != NULL && (port = atoi(preferred)) > 0) {
		/* Do nothing */
	}
	else if (   default_service != NULL
		&& (se = getservbyname(default_service, "tcp")) != NULL) {
		port = ntohs(se->s_port);
	}
	else {
		port = default_port;
	}
	return port;
}


int set_echo(int enable)
{
	struct termios t;

	if (tcgetattr(fileno(stdin), &t) < 0) {
		perror("tcgetattr()");
		return -1;
	}

	if (enable) {
		t.c_lflag |= ECHO;
	} else {
		t.c_lflag &= (~ECHO);
	}

	if (tcsetattr(fileno(stdin), TCSANOW, &t) < 0) {
		perror("tcsetattr()");
		return -1;
	}

	return 0;
}


void usage(char *progpath)
{
	printf("Usage: %s [options] [[<user>@]host[:<port>]] [..]\n"
	       "  Options:\n"
	       "    -1     : switch to version 1 (MD5) protocol\n"
	       "    -d     : enable debug messages\n"
	       "    -l user: specify user name if different from current\n"
	       "    -p pass: specify passphrase\n"
	       "    -i ip  : specity alternate hostname or IP to open\n"
	       "\n",
	       progname(progpath));
}


int connect_ipscrompd(char *host, char *dflt_user, char *password,
                      int version, char *alt_ip)
{
	char *challenge, *response,
	     *auth_str,  *at_symbol, *user = dflt_user;

	int port, auth_len, fd;

	if ((at_symbol = strchr(host, '@')) != NULL) {
		user = malloc(at_symbol - host + 1);
		if (user == NULL) {
			fprintf(stderr, "Unable to malloc() space for user string.\n");
			return 1;
		}

		strncpy(user, host, at_symbol - host);
		user[at_symbol - host] = '\0';
		host = at_symbol + 1;
	}

#ifdef __CYGWIN__
	if (user == NULL) {
		user = cuserid(NULL);
	}
#endif
	if (user == NULL) {
		fprintf(stderr, "Cannot determine username; please use -l\n");
		if (user != dflt_user) free(user);
		return 2;
	}

	port = find_port(host, DEFAULT_SERVICE, DEFAULT_PORT);

	dbg("Connecting to %s:%d\n", host, port);

	if ((fd = connect_host(host, port)) < 0) {
		if (user != dflt_user) free(user);
		return 1;
	}

	send_sock(fd, "USER %s %d\n", user, version);

	response = recv_sock(fd);

	if (response == NULL) {
		printf("Server closed connection instead of responding\n");
		if (user != dflt_user) free(user);
		close(fd);
		return 1;
	}

	if (strncmp(response, "AUTH ", 5) != 0) {
		printf("Server responded incorrectly: '%s'\n", response);
		if (user != dflt_user) free(user);
		close(fd);
		return 1;
	}

	challenge = &response[5];

	auth_len = strlen(user) + strlen(challenge) + strlen(password) + 3;
	if (alt_ip != NULL) {
		auth_len += strlen(alt_ip) + 1;
	}

	if ((auth_str = malloc(auth_len)) == NULL) {
		fprintf(stderr, "Unable to malloc() space for auth string.\n");
		if (user != dflt_user) free(user);
		close(fd);
		return 1;
	}

	if (alt_ip == NULL) {
		snprintf(auth_str, auth_len, "%s:%s:%s", user, challenge, password);
	}
	else {
		snprintf(auth_str, auth_len, "%s:%s:%s:%s", user, alt_ip, challenge, password);
	}

	if (debug > 1) {
		dbg("Auth string is: '%s'\n", auth_str);
	}

	if (alt_ip == NULL) {
		send_sock(fd, "PERMIT %s\n", hash(version, auth_str));
	}
	else {
		send_sock(fd, "IPERMIT %s %s\n", alt_ip, hash(version, auth_str));
	}

	response = recv_sock(fd);
	close(fd);

	if (user != dflt_user) free(user);

	if (strncmp(response, "OK ", 3) != 0) {
		printf("Server reports an error: '%s'\n", response);
		return 1;
	}

	printf("%s\n", response);
	return 0;
}


int main(int argc, char *argv[])
{
	int opt, version = 2, rc;
	char *user = getlogin(),
	     *pass = NULL,
	     *alt_ip = NULL, *tmp;

	while ((opt = getopt(argc, argv, "1dh:i:l:p:u:")) != EOF) {
		switch(opt) {
			case '1': version = 1;
				  break;

			case 'd': debug++;
				  break;

			case 'i': alt_ip = optarg;
				  break;

			case 'l': user = optarg;
				  break;

			case 'p': pass = optarg;
				  break;

			case '?': usage(argv[0]);
				  return 1;
				  break;

			default: fprintf(stderr, "INTERNAL ERRROR: Untrapped getopt() char '%c'\n", opt);
		}
	}

	if (alt_ip != NULL && version < 2) {
		fprintf(stderr, "WARNING: Alternative IP unsupported with old protocol\n");
	}

	if (alt_ip != NULL) {
		tmp = ip_string(alt_ip);
		if (tmp == NULL) {
			fprintf(stderr, "Cannot resolve '%s' to an IP address\n", alt_ip);
			exit(1);
		}
		alt_ip = tmp;
	}

	if (!pass) {
		char *envpass = getenv("IPSCROMP_PASS");
		if (envpass != NULL) {
			pass = envpass;
		} else {
			if (set_echo(0) < 0) {
				return 3;
			}

			pass = ask_user("Your password: ");
			printf("\n");

			/* Do we really care if this fails? What can we do? */
			set_echo(1);
		}
	}

	rc = 0;

	if (argc - optind == 0) {
		rc += connect_ipscrompd(DEFAULT_HOST, user, pass, version, alt_ip);
	}
	else {
		for (; optind < argc; optind++) {
			rc += connect_ipscrompd(argv[optind], user, pass, version, alt_ip);
		}
	}
	return rc;
}

File: ipscromp.py
#!/usr/bin/env python3
"""
ipscromp - Python client for the ipscromp firewall authentication protocol.

Usage: ipscromp.py [-1] [-d] [-l user] [-p pass] [-i alt_ip] [user@]host[:port] [...]
"""

import getpass
import hashlib
import os
import socket
import sys

DEFAULT_HOST    = "labrat.lexington.ibm.com"
DEFAULT_PORT    = 2002
DEFAULT_SERVICE = "ipscromp"
BUFFSIZE        = 4096

debug = False


def dbg(msg):
    if debug:
        print(f"DEBUG: {msg}", file=sys.stderr)


def hash_auth(version, s):
    data = s.encode()
    if version == 1:
        return hashlib.md5(data).hexdigest()
    return hashlib.sha1(data).hexdigest()


def parse_host(spec):
    """Parse [user@]host[:port].  Bare IPv6 addresses (multiple colons) are
    not split on ':' — use host:port only when there is exactly one colon."""
    user = None
    if '@' in spec:
        user, spec = spec.split('@', 1)

    first = spec.find(':')
    last  = spec.rfind(':')
    port  = None
    if first != -1 and first == last:
        host_part, port_str = spec[:first], spec[first + 1:]
        try:
            port = int(port_str)
        except ValueError:
            try:
                port = socket.getservbyname(port_str, 'tcp')
            except OSError:
                pass
        spec = host_part

    return user, spec, port


def resolve_port(port):
    if port is not None:
        return port
    try:
        return socket.getservbyname(DEFAULT_SERVICE, 'tcp')
    except OSError:
        return DEFAULT_PORT


def connect_ipscrompd(spec, default_user, password, version, alt_ip):
    user, host, port = parse_host(spec)
    if user is None:
        user = default_user
    port = resolve_port(port)

    dbg(f"connecting to {host}:{port} as {user}")

    try:
        infos = socket.getaddrinfo(host, port, socket.AF_UNSPEC, socket.SOCK_STREAM)
    except socket.gaierror as e:
        print(f"Unable to resolve '{host}': {e}", file=sys.stderr)
        return 1

    sock = None
    for af, socktype, proto, _canon, sockaddr in infos:
        dbg(f"trying {sockaddr}")
        s = None
        try:
            s = socket.socket(af, socktype, proto)
            s.connect(sockaddr)
            sock = s
            break
        except OSError:
            if s:
                s.close()

    if sock is None:
        print(f"Unable to connect to {host}:{port}", file=sys.stderr)
        return 1

    def send(msg):
        dbg(f"sending '{msg.rstrip()}'")
        sock.sendall(msg.encode())

    def recv():
        data = sock.recv(BUFFSIZE).decode().strip()
        dbg(f"received '{data}'")
        return data

    send(f"USER {user} {version}\n")

    response = recv()
    if not response.startswith("AUTH "):
        print(f"Server responded incorrectly: '{response}'")
        sock.close()
        return 1

    challenge = response[5:]

    if alt_ip is None:
        auth_str = f"{user}:{challenge}:{password}"
        h = hash_auth(version, auth_str)
        send(f"PERMIT {h}\n")
    else:
        auth_str = f"{user}:{alt_ip}:{challenge}:{password}"
        h = hash_auth(version, auth_str)
        send(f"IPERMIT {alt_ip} {h}\n")

    response = recv()
    sock.close()

    if not response.startswith("OK "):
        print(f"Server reports an error: '{response}'")
        return 1

    print(response)
    return 0


def main():
    global debug

    version     = 2
    user        = getpass.getuser()
    password    = None
    alt_ip      = None
    hosts       = []

    args = sys.argv[1:]
    i = 0
    while i < len(args):
        a = args[i]
        if a == '-1':
            version = 1
        elif a == '-d':
            debug = True
        elif a in ('-l', '-u'):
            i += 1; user = args[i]
        elif a == '-p':
            i += 1; password = args[i]
        elif a == '-i':
            i += 1; alt_ip = args[i]
        elif a in ('-h', '--help'):
            print(f"Usage: {sys.argv[0]} [-1] [-d] [-l user] [-p pass] [-i ip] "
                  f"[user@]host[:port] [...]")
            sys.exit(0)
        elif a.startswith('-'):
            print(f"Unknown option: {a}", file=sys.stderr)
            sys.exit(1)
        else:
            hosts.append(a)
        i += 1

    if alt_ip is not None and version < 2:
        print("WARNING: Alternative IP unsupported with old protocol", file=sys.stderr)

    if alt_ip is not None:
        try:
            infos = socket.getaddrinfo(alt_ip, None, socket.AF_UNSPEC)
            alt_ip = infos[0][4][0]
        except socket.gaierror:
            print(f"Cannot resolve '{alt_ip}' to an IP address", file=sys.stderr)
            sys.exit(1)

    if password is None:
        envpass = os.environ.get('IPSCROMP_PASS')
        if envpass:
            password = envpass
        else:
            password = getpass.getpass("Your password: ")

    if not hosts:
        hosts = [DEFAULT_HOST]

    rc = 0
    for host in hosts:
        rc += connect_ipscrompd(host, user, password, version, alt_ip)
    sys.exit(rc)


if __name__ == '__main__':
    main()
